Include partition key in the Iceberg data file location

mmaslankaprv opened the pull request:
Including a partition key derived path component in the Iceberg data files remote location. The partition key being part of the path allow users to better navigate the Iceberg data files in the object store. Additionally it will allow per partition access control. 

Fixes: https://redpandadata.atlassian.net/browse/CORE-8716

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
- none

mmaslankaprv committed:
datalake: include partition key in parquet file location path

Including a partition key derived path component in the Iceberg data
files remote location.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

dotnwat reviewed (commented):
> Including a partition key derived path component in the Iceberg data files remote location.

can we please state why its being done?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62390-->
<!--f033433a-9c20-40f7-8ccb-26243896f807-->
<!--0de7b6a4-aa83-4864-8ab0-15cef3d50fe7-->
<!--title-->
## Retry command for [Build#62390](https://buildkite.com/redpanda/redpanda/builds/62390)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5e57b203-fb8e-47ca-8db2-9bcb8e3b65a5-->
<!--TEST RESULTS-->
<!--fe3f217e-0eb2-4f3a-8755-a83779a3faa3-->
<!--TEST RESULTS-->
<!--3dc30f22-a23b-4e3c-831f-117597b3386a-->
<!--TEST RESULTS-->
<!--1f5a4e3d-3554-4058-ad82-914fd72d1586-->
<!--TEST RESULTS-->
<!--fa9ca06a-1ed3-45a1-961f-713c2fe157d2-->
<!--TEST RESULTS-->
<!--494ab0fa-6ba6-4bc2-9048-8c6cb802f3f3-->
<!--TEST RESULTS-->
<!--b96fe663-1b7d-4898-a41d-9c10c022954d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62390">build#62390</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62390#01954693-295b-4df0-9b2e-77fd34e611b7 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62390#01954693-295b-4df0-9b2e-77fd34e611b7 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62390#01954693-295d-4126-8c90-f1816a871e76 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62405">build#62405</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#01954796-3a9c-42f7-a483-6fcf4025e434 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#019547a9-75fc-4fab-a82c-3dafda6ae21d | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#019547a9-75fe-4431-ac8e-31d42906299a | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#019547a9-75fc-4fab-a82c-3dafda6ae21d | FAIL | 0/20 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=False.server_cfg=None.topic_cfg=90000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#019547a9-75fe-4431-ac8e-31d42906299a | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_cross_shard.num_to_upgrade=2.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#01954796-3a9d-46cf-bbd5-6feed7db1776 | FLAKY | 1/2 |
| rptest.tests.read_replica_e2e_test.TestReadReplicaService.test_simple_end_to_end.partition_count=10.cloud_storage_type_and_url_style=.CloudStorageType.S3.1.path | ducktape | https://buildkite.com/redpanda/redpanda/builds/62405#019547a9-75fe-4895-997a-c4e1325b9f5b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62463">build#62463</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionTest.test_compaction.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62463#01954de2-ab0f-438d-ace6-070e05e88a09 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62463#01954dc8-1d1b-4c27-bc2c-0eb35ed40ed9 | FAIL | 0/20 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_partition_spec_evo.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.use_partition_spec=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/62463#01954dc8-1d1c-4221-846f-3bfe14124284 | FAIL | 0/20 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62463#01954de2-ab0f-438d-ace6-070e05e88a09 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62463#01954de2-ab0e-4640-8723-efbff7ee49a6 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62541">build#62541</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62541#01955ca5-537f-47cd-8ed9-f17925a15ef3 | FLAKY | 1/2 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62541#01955ca5-537f-47d2-8dc3-87b0f48f2d58 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62661">build#62661</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| cluster_cloud_metadata_rpfixture.cluster_cloud_metadata_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/62661#019565b2-a4f0-4a33-95f8-c5a1beed76cb | FLAKY | 1/2 |
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d38-44e5-8c6b-440881585caa | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34e-4652-a942-eb140b59b660 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f350-45c8-a6bd-f4db621a2b2c | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_basic.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d39-44a1-85db-b84d2d028370 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34d-44e9-8bfd-20c5c9bdd049 | FLAKY | 1/11 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d3a-4618-a32f-84f775b5548e | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34e-4652-a942-eb140b59b660 | FLAKY | 1/3 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f350-45c8-a6bd-f4db621a2b2c | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_restore_partition_spec.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d39-44a1-85db-b84d2d028370 | FLAKY | 1/4 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34c-4ca3-b7b3-b15ad455afc7 | FLAKY | 1/8 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d39-4eb0-9492-d3c8ec0d4cc1 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34d-44e9-8bfd-20c5c9bdd049 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d3a-4618-a32f-84f775b5548e | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dlq.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34e-4652-a942-eb140b59b660 | FLAKY | 1/10 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d39-44a1-85db-b84d2d028370 | FLAKY | 1/5 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565fa-0d39-4eb0-9492-d3c8ec0d4cc1 | FLAKY | 1/2 |
| rptest.tests.datalake.cluster_restore_test.DatalakeClusterRestoreTest.test_slow_tiered_storage_dupe_records.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34d-44e9-8bfd-20c5c9bdd049 | FLAKY | 1/5 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62661#019565f6-f34c-4ca3-b7b3-b15ad455afc7 | FLAKY | 1/5 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62729">build#62729</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-30a2-4189-a60b-cc4a40274473 | FLAKY | 1/4 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e148-490d-86ef-4d1a40ea0e80 | FLAKY | 1/5 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e146-4db5-b8dd-6c76821d0dd8 | FLAKY | 1/6 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-309d-4f05-adbe-d0b9736f55a7 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e148-490d-86ef-4d1a40ea0e80 | FLAKY | 1/4 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-309f-471f-a1d8-bf89dc0873a8 | FLAKY | 1/5 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e146-4bdd-9342-f7873e7fd7a5 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-30a2-4189-a60b-cc4a40274473 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-309d-4f05-adbe-d0b9736f55a7 | FLAKY | 1/15 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e149-43e4-b395-ef1f5c2626cb | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-30a0-405c-a854-6bb284dffea9 | FLAKY | 1/5 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e148-490d-86ef-4d1a40ea0e80 | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.REST_JDBC | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-309f-471f-a1d8-bf89dc0873a8 | FLAKY | 1/2 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e149-43e4-b395-ef1f5c2626cb | FLAKY | 1/3 |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-30a0-405c-a854-6bb284dffea9 | FLAKY | 1/2 |
| rptest.tests.full_disk_test.LogStorageMaxSizeSI.test_stay_below_target_size.log_segment_size=1048576.cleanup_policy=compact | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c21-e146-4bdd-9342-f7873e7fd7a5 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62729#01956c25-30a0-405c-a854-6bb284dffea9 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62780">build#62780</a></summary>


| test_id                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_chunks           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62780#01957168-95c6-4eb6-b80b-55237e404e04 | FLAKY         | 1/2      |
| rptest.tests.consumer_group_balancing_test.ConsumerGroupBalancingTest.test_coordinator_nodes_balance | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62780#01957154-6b1e-4f96-9873-45eb750b2742 | FLAKY         | 1/2      |
| rptest.tests.mirror_maker_test.TestMirrorMakerService.test_simple_end_to_end.source_type=kafka       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62780#01957168-95c6-4be5-bc5b-8496327f9eca | FLAKY         | 1/2      |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62780#01957168-95c6-4eb6-b80b-55237e404e04 | FLAKY         | 1/2      |
</details>
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62405-->
<!--a291dc0f-c369-4868-9ce8-7e8f8b9b96cb-->
<!--f6568577-2f87-434d-a062-36321365aa0b-->
<!--title-->
## Retry command for [Build#62405](https://buildkite.com/redpanda/redpanda/builds/62405)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
```
<!--/body-->

andrwng reviewed (dismissed):
Maybe makes sense to add a simple test (ducktape or otherwise) that lists the files and ensures that they look sane? Doesn't have to be in this PR but it'd be nice as a reviewer just so we know it looks like what we expect

nvartolomei commented on src/v/datalake/translation_task.cc:
```diff
@@ -36,7 +36,7 @@ ss::future<checked<remote_path, translation_task::errc>> execute_single_upload(
   retry_chain_node& parent_rcn,
   lazy_abort_source& lazy_as) {
     auto file_remote_path = remote_path{
-      file.table_location / remote_path_prefix
```
Should we remove `remote_path_prefix` completely with this commit?

The prefix contains `n/t/p` and we were discussing removing it next time we touch this code and introduce partition key path. Once we run an optimize task with say trino/spark, they will rewrite the paths and remove that information.

nvartolomei reviewed (commented):
In addition to what Andrew said, you can try mixing together `Datalake3rdPartyMaintenanceTest` with `DatalakeE2ETests.test_iceberg_files_location` to ensure that the paths are as expected and remain so post rewrite by 3rd party engine. I.e. ensure that our interpretation of paths matches 3rd party engines*.

\* This would be best effort anyway. 3rd party query engines can be configured not to use partition key in the path, or they must even custom implementations and it is not practical to try and achieve 100% compatibility with all of them. At most, we can try to be consistent with one of them.

mmaslankaprv commented:
I've added a test.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/62463-->
<!--25a79d90-de87-492a-99bb-0b44356978f3-->
<!--09fbc2ce-eb46-4f67-9345-58235d086aa3-->
<!--title-->
## Retry command for [Build#62463](https://buildkite.com/redpanda/redpanda/builds/62463)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"trino","use_partition_spec":true}
tests/rptest/tests/datalake/schema_evolution_test.py::SchemaEvolutionE2ETests.test_partition_spec_evo@{"cloud_storage_type":1,"query_engine":"spark","use_partition_spec":true}
```
<!--/body-->

mmaslankaprv committed:
cs/s3_client: url encode target path

S3 client calculates the signature over the request transformed into a
canonical for according to the AWS specification:

https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html

It is critical for the canonical path to be exactly the same as the path
in the request otherwise the signature won't match.

This commit adds URL encoding to the S3 client in order to preserve all
already escaped characters from the original target. (The escaping of
the characters is required to meet the AWS S3 object key naming
limitations).

Previously the S3 client url encoding wasn't required and it all worked
correctly as Redpanda doesn't allow special characters in topic names.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
dl/partition_key_path: use uri encode from http utils

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
tests/dl: added test validating presence of partition key in a path

Signed-off-by: Michał Maślanka <michal@redpanda.com>

nvartolomei commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -325,6 +325,68 @@ def assert_location_prefix(rows, prefix: str):
+            spark = dl.spark()
+            table_name = f"redpanda.{self.topic_name}"
+
+            def assert_custom_partition_key_path(path):
```
I think it would be better to assert on the entire path. I.e. that is is prefixed with `<table location>` + `/data/` + `partition key as path` (split string by `/` then by `%3D` and assert at least the keys for example) + `filename`.

With this PR we are trying to match some external expectations about path formats but we don't have any test yet assertion on the actual external expectation (i.e. that parquet file are at specific locations).

Ideally we would also query the parquet files to ensure that the data is partitioned correctly and we don't write values in the wrong partition files? Or existing tests for partitioning are good enough?

(non-blocker comment)

mmaslankaprv replied:
I've changed the test to be more restrictive. I think we need an additional test of the partitioning itself

nvartolomei commented on src/v/cloud_storage_clients/s3_client.cc:
```diff
@@ -143,6 +164,7 @@ result<http::client::request_header> request_creator::make_head_object_request(
     if (ec) {
         return ec;
     }
+    url_encode_target(header);
```
did you consider putting this in `make_target`?

mmaslankaprv replied:
This has to be done post `add_auth` as it would result in double encoding when creating canonical request form for signature calculation.

bashtanov replied:
But is the encoding that happens in `split_target` needed at all if we already have it encoded in the URL?

nvartolomei commented on src/v/datalake/partition_key_path.cc:
```diff
@@ -305,7 +297,7 @@ checked<remote_path, partition_key_error> partition_key_to_path(
         }
         auto element = ssx::sformat(
-          "{}={}", element_key.value(), formatting_res.value());
+          "{}%3D{}", element_key.value(), formatting_res.value());
```
Doesn't this result in a double encoding problem if we encode equal sign here and then encode `%` again in make requests (url_encode_target)?

mmaslankaprv replied:
no, this is really confusing, i got the problem of understating that as well. The double encoding is actually required here as we want the encoded path to be received by Object Store as it requires special characters to be encoded. Not for the API but for the storage itself. 
So we want object name to contain `%3D` and in this case the `%` must be encoded when sending requests over HTTP protocol

nvartolomei replied:
I tried running the test you just added but modified it to sleep before exiting. Next I used minio client to list the keys:

```
bash-5.1# mc ls myminio/panda-bucket-b757fa9c-f8e7-11ef-a61f-0242c0a8d713/redpanda-iceberg-catalog/redpanda/test/
[2025-03-04 11:00:14 UTC]     0B metadata/
[2025-03-04 11:00:14 UTC]     0B redpanda.timestamp_hour%3D2025-03-04-10/
```

I have also used PyIceberg before with minio and here are what it's paths look like:

```
bash-5.1# mc ls myminio/localhost/warehouse/demo_table/data
[2025-02-25 11:31:57 UTC]     0B string=HjvjcmahNe/
[2025-02-25 11:31:57 UTC]     0B string=JkKbb5ZW6D/
[2025-02-25 11:31:57 UTC]     0B string=LcCjv0beCR/
```

Our path format doesn't match PyIceberg but I believe they should. Do you think otherwise?

mmaslankaprv replied:
i read the naming doc once again and i think that we can use `=` sign instead of `%3D`

mmaslankaprv replied:
i will change the ordering as well

andrwng commented on src/v/datalake/translation/partition_translator.cc:
```diff
@@ -303,8 +303,6 @@ partition_translator::do_translation_for_range(
       _location_provider,
       *_probe,
     };
-    auto remote_path_prefix = remote_path{
```
I agree with folks saying we shouldn't depend on this because external writers can overwrite these files. But what's the harm in having them, if anything even as a part of the filename? They seem potentially useful for debugging.

mmaslankaprv replied:
I am happy to drop that change, we can always do it later

mmaslankaprv replied:
@nvartolomei can you chime in as you requested me to remove the `ntp` component from the path

nvartolomei replied:
* Moving it to file name—ok with doing that.
* With s3 I’m not as concerned about it being in the path (before the last slash) but even there users might have expectations that it’s a flat structure under partition and use list prefix with a separator which has a different api response if there is additional hierarchy.
* With abs I’m a bit more concerned, it is an actual directory structure and the expectations might be that there is no additional nesting.
* We are guessing on coffee grounds here so I’m fine with a common denominator. Also more info is always better. Moving it to the file name is perfect.

andrwng commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -325,6 +327,78 @@ def assert_location_prefix(rows, prefix: str):
+
+                path_parts = m['key'].split("/")
+                partition_key = path_parts[3]
+                dt = datetime.datetime.now()
```
Shouldn't this be something like, the `datetime.datetime.now()` from right after calling our first produce? Otherwise it seems like this could be flaky when the year changes :)

mmaslankaprv replied:
yeah, this is not perfect around the new years eve

andrwng reviewed (commented):
Thanks for adding the test! Also had one question about the naming change

nvartolomei reviewed (dismissed):
LGTM! I'd like to see a cdt run for datalake tests with azure before merging.

mmaslankaprv committed:
dl/manager: remove term and partition from path

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
/cdt  
provider=azure  
region=centralus  
cdt_instance_type=Standard_L8as_v3  
tests/rptest/tests/datalake/datalake_e2e_test.py

ztlpn commented on tests/rptest/tests/datalake/datalake_e2e_test.py:
```diff
@@ -332,6 +334,78 @@ def assert_location_prefix(rows, prefix: str):
+    @cluster(num_nodes=3)
+    @matrix(cloud_storage_type=supported_storage_types(),
+            catalog_type=supported_catalog_types(),
+            custom_partition_spec=[None, "(timestamp_us)", "(number)"])
```
should we test with a multi-field spec as well?

mmaslankaprv replied:
I think we need a separate test that will test custom partitions end to end

ztlpn replied:
It is an interesting test case for file locations as well, right? Verifying the `/field1=val1/field2=val2/...` structure.

ztlpn replied:
> we need a separate test that will test custom partitions end to end

There is rptest.tests.datalake.custom_partitioning_test.DatalakeCustomPartitioningTest.test_basic

mmaslankaprv committed:
cs/abs_client: added url encoding for paths

Added URL encoding of HTTP paths. The encoding is required right now as
the paths may contain special characters coming from the Iceberg
partition key.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv commented:
Latest force push: 
- added URL encoding in ABS client

ztlpn commented on src/v/datalake/translation_task.cc:
```diff
@@ -36,7 +36,7 @@ ss::future<checked<remote_path, translation_task::errc>> execute_single_upload(
     auto file_remote_path = remote_path{
-      file.table_location / remote_path_prefix
+      file.table_location / remote_path_prefix / file.partition_key_path
       / file.local_file.path().filename()};
```
Probably unrelated to this PR, but I see that the parquet files are generated with a weird name starting with `-` (e.g. `-7c0a39de-41d7-4531-80d0-ebfb40e054c3.parquet`).

nvartolomei replied:
This was "broken" in https://github.com/redpanda-data/redpanda/pull/25152/files#diff-617058a8220c70fdc909b169e594470f7b81ed70497f71c95f421c9da374aee8L290

vs 

https://github.com/redpanda-data/redpanda/blob/502834a7ee95102e1280a486b2c386a9656a08cd/src/v/datalake/translation/deps.cc#L482

ztlpn replied:
Per an offline discussion with Michal, he'll fix this (and add term/partition-id to the filename) in a followup.

mmaslankaprv commented:
forced pushed to resolve conflict
